src/ngraph/pass/constant_folding_non_zero.cpp

```diff
@@ -0,0 +1,198 @@
+//*****************************************************************************
+// Copyright 2017-2020 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//*****************************************************************************
+
+#include "constant_folding.hpp"
+#include "ngraph/op/non_zero.hpp"
+
+using namespace std;
+using namespace ngraph;
+
+namespace
+{
+    struct NonZeroElements
+    {
+        using Results_t = std::vector<std::vector<int64_t>>;
+
+        NonZeroElements(const Shape& input_shape)
+            : m_input_shape{input_shape}
+            , m_results{Results_t(input_shape.size())}
+        {
+        }
+
+        template <typename T>
+        const Results_t& find_indices(const T* values)
+        {
+            m_current_index = Shape(m_input_shape.size(), 0UL);
```

Comment:
There is a trivial function Constant::get_all_data_elements_bitwise_identical() which returns if all elements are identical. This is computed as the constant is constructed so it simply returns a stored value. You should check if you need to go any further in this routine since it iterates through all values of every constant in a model which is expensive.

Reply:
Thanks for this hint, I've used this method to short circuit here https://github.com/NervanaSystems/ngraph/pull/4431/files?file-filters%5B%5D=.cpp#diff-7079b532eda2d11cc6adcf43a0a6d972R119-R124